spec/lib/s3_cors_rulesets_spec.rb

```diff
@@ -0,0 +1,209 @@
+      client.expects(:get_bucket_cors).never
+      sync_rules
+    end
+
+    context "when S3 is set up with global settings" do
+      let(:use_db_s3_config) { false }
+      before do
+        global_setting :s3_use_iam_profile, true
+        global_setting :s3_bucket, "s3-upload-bucket"
+        global_setting :s3_backup_bucket, "s3-backup-upload-bucket"
+        global_setting :s3_region, "us-west-2"
+      end
+
+      it "does nothing if !s3_install_cors_rule" do
+        SiteSetting.s3_install_cors_rule = false
+        sync_rules
+      end
+
+      it "only tries to apply the ASSETS rules by default" do
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        sync_rules
+      end
+
+      it "does not apply the ASSETS rules if they already exist" do
+        client.stub_responses(:get_bucket_cors, {
+          cors_rules: [S3CorsRulesets::ASSETS]
+        })
+        client.expects(:put_bucket_cors).never
+        sync_rules
+      end
+
+      it "applies the ASSETS rules and the BACKUP_DIRECT_UPLOAD rules if S3 backups are enabled" do
+        setup_backups
+
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-backup-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::BACKUP_DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(false)
+        expect(result[:backup_rules_applied]).to eq(true)
+      end
+
+      it "applies the ASSETS rules and the DIRECT_UPLOAD rules when S3 direct uploads are enabled" do
+        SiteSetting.enable_direct_s3_uploads = true
+
+        client.stub_responses(:get_bucket_cors, {})
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::ASSETS
+            ]
+          }
+        )
+        client.expects(:put_bucket_cors).with(
+          bucket: "s3-upload-bucket",
+          cors_configuration: {
+            cors_rules: [
+              S3CorsRulesets::DIRECT_UPLOAD
+            ]
+          }
+        )
+        result = sync_rules
+        expect(result[:assets_rules_applied]).to eq(true)
+        expect(result[:direct_upload_rules_applied]).to eq(true)
+        expect(result[:backup_rules_applied]).to eq(false)
+      end
```

Comment:
I kind of feel like all these tests are not necessary since they are eventually covered by the "applies all rules if everything is set up" test below.

Reply:
I would prefer to get rid of the "applies all rules" test then, I want to have explicit tests for each of the "configured" rules cause they have distinct logic.